File: web/src/engine/levels.ts
/**
 * The two knobs a player gets, and why they are two rather than one.
 *
 * **Which generation** sets how good the intuition is. These are real
 * checkpoints from the training run, and the numbers attached to them come from
 * the cross-generation tournament -- 210 games per entrant, Bradley-Terry
 * ratings with bootstrap intervals, anchored at random play = 0. They are not
 * adjectives.
 *
 * **How many simulations** sets how hard it thinks. The same network searching
 * 800 positions plays noticeably better than one searching 16, which is the
 * whole idea behind the method: the search improves on the network's first
 * guess, and more search improves on it more.
 *
 * Keeping them separate means a beginner can play the final agent thinking
 * briefly, or an early agent thinking hard, and the difference between those two
 * is the thing this project is actually about.
 *
 * A note on the third control, the value guardrail. At the easier levels the
 * agent samples among reasonable moves rather than always taking the best one.
 * The guardrail throws away candidates whose value is much worse than the best
 * before sampling, which is what makes an easy opponent *weak* rather than
 * *stupid*: it plays a second-best move instead of giving away a corner. Filter
 * first, then sample -- the other order lets a blunder through whenever it
 * happens to be picked.
 */

export interface Level {
  readonly id: string;
  readonly label: string;
  readonly simulations: number;
  /** 0 plays the most-visited move; above 0 samples in proportion to visits. */
  readonly temperature: number;
  /** Sample among at most this many candidates, after the guardrail. */
  readonly topK?: number;
  /** Discard candidates whose value is worse than the best by more than this. */
  readonly guard: number;
  /**
   * Stop searching after this long, whatever the simulation count says.
   *
   * Measured in a browser on a 20-core laptop at four threads: 16 simulations
   * takes 55 ms, 64 takes 221 ms, 256 takes 898 ms and 800 takes 2.8 seconds.
   * The last of those is past the point where a move reads as thinking rather
   * than as a hang, and a phone would be several times slower again.
   *
   * So the deep levels are capped by time. The count becomes a ceiling and the
   * clock decides, which means the same level feels the same everywhere and
   * gets stronger on better hardware instead of slower on worse.
   */
  readonly maxMillis?: number;
  readonly description: string;
}

export const LEVELS: readonly Level[] = [
  {
    id: "casual",
    label: "Casual",
    simulations: 16,
    temperature: 0.8,
    topK: 3,
    guard: 0.35,
    description: "Looks a little way ahead and does not always take its best move.",
  },
  {
    id: "club",
    label: "Club",
    simulations: 64,
    temperature: 0.35,
    topK: 2,
    guard: 0.2,
    description: "Considers more, and mostly plays what it considers best.",
  },
  {
    id: "strong",
    label: "Strong",
    simulations: 256,
    temperature: 0,
    guard: 0.05,
    maxMillis: 1200,
    description: "Plays its best move every time. About a second per move.",
  },
  {
    id: "max",
    label: "Max",
    simulations: 800,
    temperature: 0,
    guard: 0,
    maxMillis: 2000,
    description:
      "Thinks for up to two seconds, and searches as far as your device allows in that time.",
  },
];

export function levelById(id: string): Level {
  const found = LEVELS.find((level) => level.id === id);
  if (found === undefined) {
    throw new Error(`unknown difficulty ${id}; expected one of ${LEVELS.map((l) => l.id).join(", ")}`);
  }
  return found;
}

/**
 * Choose which move to actually play, given what the search found.
 *
 * The guardrail is applied *before* sampling, never after. Sampling first and
 * then checking would mean a genuinely bad move gets played whenever it happens
 * to be drawn, which is exactly the behaviour the guardrail exists to prevent.
 */
export function chooseMove(
  result: { actions: number[]; visits: number[]; qValues: number[] },
  level: Level,
  random: () => number = Math.random,
): number {
  const { actions, visits, qValues } = result;
  if (actions.length === 0) throw new Error("no actions to choose from");
  if (actions.length === 1) return actions[0]!;

  let best = 0;
  for (let i = 1; i < visits.length; i++) if (visits[i]! > visits[best]!) best = i;
  if (level.temperature <= 0) return actions[best]!;

  // Step 1: throw away anything clearly worse than the best move available.
  const bestQ = Math.max(...qValues);
  let candidates = actions
    .map((action, index) => ({ action, visits: visits[index]!, q: qValues[index]! }))
    .filter((candidate) => bestQ - candidate.q <= level.guard);

  if (candidates.length === 0) return actions[best]!;

  // Step 2: keep the most-searched few of what survived.
  candidates.sort((a, b) => b.visits - a.visits);
  if (level.topK !== undefined) candidates = candidates.slice(0, level.topK);

  // Step 3: sample among those, sharpened by the temperature.
  const weights = candidates.map((candidate) =>
    Math.pow(Math.max(candidate.visits, 0), 1 / level.temperature),
  );
  const total = weights.reduce((a, b) => a + b, 0);
  if (!Number.isFinite(total) || total <= 0) return candidates[0]!.action;

  let draw = random() * total;
  for (let i = 0; i < candidates.length; i++) {
    draw -= weights[i]!;
    if (draw <= 0) return candidates[i]!.action;
  }
  return candidates[candidates.length - 1]!.action;
}

File: src/reversi/difficulty/levels.py
"""Four difficulty levels from one network, and the trick that makes the easy
ones pleasant to play against.

**The obvious way to make an agent weaker is to make it random, and it is wrong.**
An agent that plays randomly some fraction of the time does not feel like a
weaker opponent; it feels like a broken one. It will hand you a corner for no
reason, then play three strong moves, then blunder its whole position. Losing to
it is annoying and beating it teaches you nothing.

So weakness here comes from three dials, and one guardrail:

* **How long it thinks.** 16 simulations against 800 is the biggest single lever
  on strength, and the most honest: the agent is genuinely considering less.
* **How much it samples.** At temperature 0 it always plays its best move. Above
  0 it picks in proportion to how much it liked each one, so it varies.
* **How many moves it will consider.** Restricting to the top few visits keeps it
  from picking something it barely looked at.

* **The guardrail** is what stops sampling from producing nonsense. After the
  search, any move whose value is worse than the best by more than ``guard`` is
  discarded *before* sampling. So an easy level plays a move it thinks is
  slightly worse -- never one it knows is terrible. It gives away small
  advantages, not corners.

That is the difference between an opponent that is beatable and one that is
broken, and it is the reason ``guard`` exists at all.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from reversi.errors import ConfigError
from reversi.search.mcts import SearchResult
from reversi.types import Action

__all__ = ["LEVELS", "DifficultyLevel", "choose_move", "level_by_name"]


@dataclass(frozen=True, slots=True)
class DifficultyLevel:
    """One rung of the ladder."""

    name: str
    label: str
    simulations: int
    temperature: float
    top_k: int | None
    guard: float
    """How far below the best move's value a candidate may be and still be
    played. 0 means "only the best move"; 0.35 means "anything not clearly
    losing"."""

    description: str = ""

    def __post_init__(self) -> None:
        if self.simulations < 1:
            msg = f"{self.name}: simulations must be at least 1"
            raise ConfigError(msg)
        if self.temperature < 0.0:
            msg = f"{self.name}: temperature cannot be negative"
            raise ConfigError(msg)
        if self.top_k is not None and self.top_k < 1:
            msg = f"{self.name}: top_k must be at least 1 when set"
            raise ConfigError(msg)
        if self.guard < 0.0:
            msg = f"{self.name}: guard cannot be negative"
            raise ConfigError(msg)


LEVELS: tuple[DifficultyLevel, ...] = (
    DifficultyLevel(
        name="casual",
        label="Casual",
        simulations=16,
        temperature=0.8,
        top_k=3,
        guard=0.35,
        description="Thinks briefly and varies its play. Gives away small advantages.",
    ),
    DifficultyLevel(
        name="club",
        label="Club",
        simulations=64,
        temperature=0.35,
        top_k=2,
        guard=0.20,
        description="Considers more, and mostly plays what it considers best.",
    ),
    DifficultyLevel(
        name="strong",
        label="Strong",
        simulations=256,
        temperature=0.0,
        top_k=None,
        guard=0.05,
        description="Plays its best move every time.",
    ),
    DifficultyLevel(
        name="max",
        label="Max",
        simulations=800,
        temperature=0.0,
        top_k=None,
        guard=0.0,
        description="Thinks as long as it is allowed to, and never compromises.",
    ),
)


def level_by_name(name: str) -> DifficultyLevel:
    for level in LEVELS:
        if level.name == name:
            return level
    known = ", ".join(level.name for level in LEVELS)
    msg = f"unknown difficulty {name!r}; expected one of {known}"
    raise ConfigError(msg)


def choose_move(
    result: SearchResult,
    level: DifficultyLevel,
    rng: np.random.Generator | None = None,
) -> Action:
    """Pick a move at this difficulty, applying the guardrail first.

    Order matters: **filter, then sample.** Sampling first and filtering after
    would let a bad move be chosen and then vetoed, which changes the
    distribution in a way that is hard to reason about. Filtering first means the
    agent is choosing among moves it considers acceptable, which is exactly what
    a weaker but sane player does.
    """
    if not result.actions:
        msg = "cannot choose a move from a search with no actions"
        raise ConfigError(msg)

    candidates = _acceptable(result, level)

    if level.temperature <= 0.0 or len(candidates) == 1:
        # Best of what survived the guard -- most visited, as always.
        best = max(candidates, key=lambda i: result.visits[i])
        return result.actions[best]

    if rng is None:
        msg = f"difficulty {level.name!r} samples its move, so it needs an rng"
        raise ConfigError(msg)

    weights = np.array([result.visits[i] for i in candidates], dtype=np.float64)
    if weights.sum() <= 0:
        return result.actions[candidates[0]]

    sharpened = weights ** (1.0 / level.temperature)
    total = sharpened.sum()
    if not np.isfinite(total) or total <= 0.0:
        best = max(candidates, key=lambda i: result.visits[i])
        return result.actions[best]

    chosen = int(rng.choice(len(candidates), p=sharpened / total))
    return result.actions[candidates[chosen]]


def _acceptable(result: SearchResult, level: DifficultyLevel) -> list[int]:
    """Indices of the moves this level is willing to play.

    Two filters, both against the *best* move rather than an absolute threshold:
    a position where every move is bad should still produce a move.
    """
    order = sorted(range(len(result.actions)), key=lambda i: -result.visits[i])

    # The guardrail runs first, and top_k narrows what survives it.
    #
    # The other order looks equivalent and is not. Taking the three most-visited
    # moves first makes the guard compare against the best of *those three*
    # rather than the best move available, so a strong move that happened to be
    # searched less becomes invisible and a weaker one passes a guard it should
    # have failed. The criterion is worded against the best move available, and
    # this is the order that delivers that.
    #
    # Found by the calibration: one violation in 500 moves, which is exactly the
    # rate you would expect from a case this narrow.
    if level.guard > 0.0:
        # Only judge moves the search actually looked at -- an unvisited move has
        # no measured value, and treating its 0.0 as an opinion would let a level
        # with a wide guard play something it never considered.
        visited = [i for i in order if result.visits[i] > 0]
        if visited:
            best_value = max(result.q_values[i] for i in visited)
            order = [i for i in visited if result.q_values[i] >= best_value - level.guard]

    if level.top_k is not None:
        order = order[: level.top_k]

    return order or [max(range(len(result.actions)), key=lambda i: result.visits[i])]
